File: packages/types/src/billing.ts
/**
 * Copyright (c) 2023-present Plane Software, Inc. and contributors
 * SPDX-License-Identifier: AGPL-3.0-only
 * See the LICENSE file for details.
 *
 * GAM addition: Customer/Service billing types.
 */

// GAM: monthly services are one fixed line a month; per-job services are billed per completed item
export type TServiceBillingType = "per_job" | "monthly";

// GAM: language of everything the client receives
export type TCustomerLanguage = "el" | "en";

export interface IServiceTemplateItem {
  id: string;
  service_id: string;
  name: string;
  sequence: number;
}

export interface IService {
  id: string;
  workspace_id: string;
  name: string;
  billing_type: TServiceBillingType;
  is_active: boolean;
  template_items: IServiceTemplateItem[];
}

export interface ICustomerServiceRate {
  id: string;
  customer_id: string;
  service: string;
  service_name: string;
  service_billing_type: TServiceBillingType;
  price: string;
}

export interface ICustomer {
  id: string;
  workspace_id: string;
  name: string;
  contact_email: string;
  language: TCustomerLanguage;
  is_active: boolean;
  rates: ICustomerServiceRate[];
}

export interface IIssueCustomerService {
  id: string;
  issue: string;
  customer: string;
  customer_name: string;
  service: string;
  service_name: string;
}

// GAM addition: repeating work items
export type TIssueRecurrenceFrequency = "weekly" | "monthly" | "yearly";

export interface IIssueRecurrence {
  id: string;
  issue: string;
  frequency: TIssueRecurrenceFrequency;
  next_run_date: string;
}

File: apps/api/plane/app/serializers/billing.py
# Copyright (c) 2023-present Plane Software, Inc. and contributors
# SPDX-License-Identifier: AGPL-3.0-only
# See the LICENSE file for details.
#
# GAM addition: serializers for the Customer/Service billing system.

from rest_framework import serializers

from plane.db.models import Customer, CustomerServiceRate, IssueCustomerService, Service, ServiceTemplateItem

from .base import BaseSerializer


class ServiceTemplateItemSerializer(BaseSerializer):
    class Meta:
        model = ServiceTemplateItem
        fields = ["id", "service_id", "name", "sequence"]
        read_only_fields = ["service"]


class ServiceSerializer(BaseSerializer):
    template_items = ServiceTemplateItemSerializer(many=True, read_only=True)

    class Meta:
        model = Service
        fields = ["id", "workspace_id", "name", "is_active", "template_items"]
        read_only_fields = ["workspace"]


class CustomerServiceRateSerializer(BaseSerializer):
    service_name = serializers.CharField(source="service.name", read_only=True)

    class Meta:
        model = CustomerServiceRate
        fields = ["id", "customer_id", "service", "service_name", "price"]
        read_only_fields = ["customer"]


class CustomerSerializer(BaseSerializer):
    rates = CustomerServiceRateSerializer(many=True, read_only=True)

    class Meta:
        model = Customer
        fields = [
            "id",
            "workspace_id",
            "name",
            "contact_email",
            "billing_type",
            "is_active",
            "rates",
        ]
        read_only_fields = ["workspace"]


class IssueCustomerServiceSerializer(BaseSerializer):
    customer_name = serializers.CharField(source="customer.name", read_only=True)
    service_name = serializers.CharField(source="service.name", read_only=True)

    class Meta:
        model = IssueCustomerService
        fields = ["id", "issue", "customer", "customer_name", "service", "service_name"]
        read_only_fields = ["issue"]
